File: src/cleaning/data_cleaner.py
"""
Limpeza e padronização de dados.

Regras de negócio (não negociáveis, aprendidas de erros anteriores do projeto):
  - Nunca aplicar a mesma estratégia de nulo para todas as colunas de uma vez.
  - Sempre medir e logar quantas linhas/valores foram afetados por cada etapa —
    isso é o que permite auditar e defender o resultado depois.
  - Nunca sobrescrever o DataFrame original sem gerar um retorno novo — quem
    chama decide se quer substituir ou manter os dois para comparar.
"""

import pandas as pd


class DataCleaner:

    def __init__(self, logger=None):
        self.logger = logger
        self.resumo_execucao = []

    def _log(self, mensagem: str):
        self.resumo_execucao.append(mensagem)
        if self.logger:
            self.logger.info(mensagem)
        else:
            print(f"[LIMPEZA] {mensagem}")

    def padronizar_texto(self, df: pd.DataFrame, colunas: list) -> pd.DataFrame:
        """Remove espaços extras e padroniza caixa em colunas categóricas de texto."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' não encontrada — pulando padronização de texto.")
                continue
            antes = df[col].nunique(dropna=True)
            df[col] = df[col].astype(str).str.strip().str.upper()
            depois = df[col].nunique(dropna=True)
            self._log(
                f"Coluna '{col}': padronizada (strip + upper). "
                f"Categorias distintas antes={antes}, depois={depois}."
            )
        return df

    def tratar_nulos(self, df: pd.DataFrame, estrategia_por_coluna: dict) -> pd.DataFrame:
        """
        estrategia_por_coluna: {"coluna": "mediana" | "moda" | "remover_linha" | valor_fixo}
        """
        df = df.copy()
        for col, estrategia in estrategia_por_coluna.items():
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' não encontrada — pulando tratamento de nulo.")
                continue

            nulos_antes = df[col].isna().sum()
            if nulos_antes == 0:
                continue

            if estrategia == "mediana":
                df[col] = df[col].fillna(df[col].median())
            elif estrategia == "moda":
                moda = df[col].mode()
                if len(moda) > 0:
                    df[col] = df[col].fillna(moda[0])
            elif estrategia == "remover_linha":
                total_antes = len(df)
                df = df.dropna(subset=[col])
                removidas = total_antes - len(df)
                self._log(
                    f"Coluna '{col}': removidas {removidas} linhas com valor nulo "
                    f"({removidas / total_antes * 100:.1f}% da base)."
                )
                continue
            else:
                df[col] = df[col].fillna(estrategia)

            self._log(f"Coluna '{col}': {nulos_antes} valores nulos tratados (estratégia: '{estrategia}').")
        return df

    def remover_duplicatas(self, df: pd.DataFrame, subset: list = None) -> pd.DataFrame:
        total_antes = len(df)
        df = df.drop_duplicates(subset=subset)
        removidas = total_antes - len(df)
        pct = (removidas / total_antes * 100) if total_antes > 0 else 0
        self._log(f"Duplicatas removidas: {removidas} ({pct:.2f}% da base).")
        return df

    def corrigir_tipos(self, df: pd.DataFrame, tipos_esperados: dict) -> pd.DataFrame:
        """
        tipos_esperados: {"coluna": "numero" | "texto" | "data"}
        Valores que não conseguem ser convertidos viram NaN (não travam a execução).
        """
        df = df.copy()
        for col, tipo in tipos_esperados.items():
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' não encontrada — pulando conversão de tipo.")
                continue

            if tipo == "numero":
                antes_invalidos = pd.to_numeric(df[col], errors="coerce").isna().sum()
                df[col] = pd.to_numeric(df[col], errors="coerce")
                self._log(f"Coluna '{col}': convertida para número. {antes_invalidos} valores não numéricos viraram nulo.")
            elif tipo == "data":
                df[col] = pd.to_datetime(df[col], errors="coerce")
                self._log(f"Coluna '{col}': convertida para data.")
            elif tipo == "texto":
                df[col] = df[col].astype(str)
                self._log(f"Coluna '{col}': convertida para texto.")
        return df

    def detectar_outliers_iqr(self, df: pd.DataFrame, coluna: str) -> pd.DataFrame:
        """
        Retorna apenas as linhas identificadas como outlier (método IQR),
        para inspeção manual — não remove nada automaticamente, porque
        outlier pode ser dado real e essa decisão exige julgamento humano.
        """
        if coluna not in df.columns:
            self._log(f"Aviso: coluna '{coluna}' não encontrada — não é possível checar outliers.")
            return pd.DataFrame()

        q1 = df[coluna].quantile(0.25)
        q3 = df[coluna].quantile(0.75)
        iqr = q3 - q1
        limite_inferior = q1 - 1.5 * iqr
        limite_superior = q3 + 1.5 * iqr

        outliers = df[(df[coluna] < limite_inferior) | (df[coluna] > limite_superior)]
        self._log(
            f"Coluna '{coluna}': {len(outliers)} possíveis outliers encontrados "
            f"(fora do intervalo [{limite_inferior:.2f}, {limite_superior:.2f}]). "
            "Revisar manualmente antes de decidir remover."
        )
        return outliers

    def substituir_valores(self, df: pd.DataFrame, colunas: list,
                           mapeamento: dict) -> pd.DataFrame:
        """Substitui valores exatos conforme mapeamento {antigo: novo} em colunas selecionadas."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada — pulando substituicao.")
                continue
            antes_valores = df[col].value_counts().to_dict()
            df[col] = df[col].replace(mapeamento)
            depois_valores = df[col].value_counts().to_dict()
            total_trocas = sum(antes_valores.get(k, 0) for k in mapeamento if k in antes_valores)
            self._log(
                f"Coluna '{col}': {total_trocas} valores substituidos. "
                f"Mapeamento: {mapeamento}"
            )
        return df

    def padronizar_texto_minusculas(self, df: pd.DataFrame, colunas: list) -> pd.DataFrame:
        """Converte texto para minusculo (diferente de padronizar_texto que usa UPPER)."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada.")
                continue
            df[col] = df[col].astype(str).str.strip().str.lower()
            self._log(f"Coluna '{col}': convertida para minusculo.")
        return df

    def extrair_texto_regex(self, df: pd.DataFrame, colunas: list,
                            padrao: str, nome_nova_coluna: str = None) -> pd.DataFrame:
        """Extrai partes de texto de uma coluna usando regex e cria nova coluna."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada.")
                continue
            saida = nome_nova_coluna or f"{col}_extraido"
            df[saida] = df[col].astype(str).str.extract(padrao, expand=False)
            self._log(f"Regex '{padrao}' aplicada na coluna '{col}' -> '{saida}'.")
        return df

    def tratar_nulos_por_colunas(self, df: pd.DataFrame, colunas: list,
                                 valor_padrao) -> pd.DataFrame:
        """Preenche nulos com um valor fixo especifico (string, numero, etc)."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada.")
                continue
            nulos = df[col].isna().sum()
            if nulos > 0:
                df[col] = df[col].fillna(valor_padrao)
                self._log(f"Coluna '{col}': {nulos} nulos preenchidos com '{valor_padrao}'.")
        return df

    def padronizar_datas(self, df: pd.DataFrame, colunas: list,
                         formato_saida: str = "%d/%m/%Y") -> pd.DataFrame:
        """Converte colunas de texto/data para formato padronizado."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada.")
                continue
            try:
                df[col] = pd.to_datetime(df[col], errors="coerce")
                formato_br = df[col].dt.strftime(formato_saida)
                df[col] = formato_br
                self._log(f"Coluna '{col}': datas padronizadas no formato {formato_saida}.")
            except Exception as e:
                self._log(f"ERRO ao padronizar datas na coluna '{col}': {e}")
        return df

    def normalizar_numeros(self, df: pd.DataFrame, colunas: list,
                           casas_decimais: int = 2) -> pd.DataFrame:
        """Arredonda valores numericos para o numero especificado de casas decimais."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada.")
                continue
            df[col] = pd.to_numeric(df[col], errors="coerce")
            df[col] = df[col].round(casas_decimais)
            self._log(f"Coluna '{col}': numeros arredondados para {casas_decimais} casas.")
        return df

    def substituir_regex(self, df: pd.DataFrame, colunas: list,
                         padrao: str, novo: str) -> pd.DataFrame:
        """Substitui texto em colunas usando expressao regular."""
        df = df.copy()
        for col in colunas:
            if col not in df.columns:
                self._log(f"Aviso: coluna '{col}' nao encontrada — pulando substituicao regex.")
                continue
            antes = df[col].nunique(dropna=True)
            df[col] = df[col].astype(str).str.replace(padrao, novo, regex=True)
            depois = df[col].nunique(dropna=True)
            self._log(
                f"Regex '{padrao}' -> '{novo}' aplicada em '{col}'. "
                f"Categorias antes={antes}, depois={depois}."
            )
        return df

    def remover_colunas_vazias(self, df: pd.DataFrame, limiar: float = 1.0) -> pd.DataFrame:
        """Remove colunas onde todos (ou quase todos) os valores sao nulos."""
        df = df.copy()
        limiar_nulos = int(len(df) * limiar)
        colunas_vazias = [c for c in df.columns if df[c].isna().sum() >= limiar_nulos]
        if colunas_vazias:
            df = df.drop(columns=colunas_vazias)
            self._log(f"Colunas vazias removidas ({limiar*100:.0f}% nulos): {colunas_vazias}")
        else:
            self._log("Nenhuma coluna totalmente vazia encontrada.")
        return df

    def relatorio_resumo(self, df_antes: pd.DataFrame, df_depois: pd.DataFrame) -> str:
        linhas_removidas = len(df_antes) - len(df_depois)
        pct = (linhas_removidas / len(df_antes) * 100) if len(df_antes) > 0 else 0
        resumo = (
            f"Linhas: {len(df_antes)} -> {len(df_depois)} ({linhas_removidas} removidas, {pct:.1f}%) | "
            f"Colunas: {df_antes.shape[1]} -> {df_depois.shape[1]}"
        )
        self._log(f"RESUMO FINAL: {resumo}")
        return resumo

File: src/cleaning/dataCleaner.ts
import { DataFrame, DataRow } from "../types";
import { Logger } from "../utils/logger";

export class DataCleaner {
  private logger?: Logger;

  constructor(logger?: Logger) {
    this.logger = logger;
  }

  private _log(msg: string) {
    if (this.logger) {
      this.logger.info(msg);
    }
  }

  private isNull(val: any): boolean {
    return val === null || val === undefined || val === "" || (typeof val === "number" && isNaN(val));
  }

  /**
   * Remove colunas onde proporção de nulos >= limiar (padrão 1.0 = 100% vazia).
   */
  public removerColunasVazias(df: DataFrame, limiar: number = 1.0): DataFrame {
    const totalLinhas = df.data.length;
    if (totalLinhas === 0) return df;

    const colunasManter: string[] = [];
    const colunasRemovidas: string[] = [];

    for (const col of df.columns) {
      let nulos = 0;
      for (const row of df.data) {
        if (this.isNull(row[col])) nulos++;
      }
      const propNulos = nulos / totalLinhas;
      if (propNulos >= limiar) {
        colunasRemovidas.push(col);
      } else {
        colunasManter.push(col);
      }
    }

    const newData = df.data.map((row) => {
      const newRow: DataRow = {};
      for (const col of colunasManter) {
        newRow[col] = row[col];
      }
      return newRow;
    });

    this._log(`[COLUNAS_VAZIAS] ${colunasRemovidas.length} colunas removidas: ${colunasRemovidas.join(", ")}`);

    return {
      columns: colunasManter,
      data: newData,
    };
  }

  /**
   * Remove linhas vazias (todas ou qualquer coluna nula).
   */
  public removerLinhasVazias(
    df: DataFrame,
    colunas?: string[],
    modo: "todas" | "qualquer" = "todas"
  ): DataFrame {
    const checkCols = colunas && colunas.length > 0 ? colunas.filter((c) => df.columns.includes(c)) : df.columns;
    if (checkCols.length === 0) return df;

    const filteredData = df.data.filter((row) => {
      if (modo === "todas") {
        return !checkCols.every((col) => this.isNull(row[col]));
      } else {
        return !checkCols.some((col) => this.isNull(row[col]));
      }
    });

    const removidas = df.data.length - filteredData.length;
    this._log(`[LINHAS_VAZIAS] ${removidas} linhas removidas (${df.data.length} -> ${filteredData.length}).`);

    return {
      columns: [...df.columns],
      data: filteredData,
    };
  }

  /**
   * Trata valores nulos com diferentes estratégias.
   */
  public tratarNulos(
    df: DataFrame,
    colunas: string[],
    estrategia: "mediana" | "media" | "moda" | "valor_fixo" | "remover_linha" | "ffill" | "bfill" = "valor_fixo",
    valorFixo: any = ""
  ): DataFrame {
    let currentData = df.data.map((r) => ({ ...r }));

    if (estrategia === "remover_linha") {
      const filtered = currentData.filter((row) => {
        return !colunas.some((col) => df.columns.includes(col) && this.isNull(row[col]));
      });
      this._log(`[NULOS] ${currentData.length - filtered.length} linhas com nulos removidas.`);
      return { columns: [...df.columns], data: filtered };
    }

    for (const col of colunas) {
      if (!df.columns.includes(col)) continue;

      if (estrategia === "valor_fixo") {
        for (const row of currentData) {
          if (this.isNull(row[col])) {
            row[col] = valorFixo;
          }
        }
      } else if (estrategia === "media" || estrategia === "mediana") {
        const nums: number[] = [];
        for (const row of currentData) {
          if (!this.isNull(row[col])) {
            const n = parseFloat(String(row[col]).replace(/\./g, "").replace(",", "."));
            if (!isNaN(n)) nums.push(n);
          }
        }

        if (nums.length > 0) {
          let repVal: number;
          if (estrategia === "media") {
            repVal = nums.reduce((a, b) => a + b, 0) / nums.length;
          } else {
            nums.sort((a, b) => a - b);
            const mid = Math.floor(nums.length / 2);
            repVal = nums.length % 2 !== 0 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2;
          }
          for (const row of currentData) {
            if (this.isNull(row[col])) {
              row[col] = Number(repVal.toFixed(2));
            }
          }
        }
      } else if (estrategia === "moda") {
        const counts = new Map<any, number>();
        for (const row of currentData) {
          if (!this.isNull(row[col])) {
            counts.set(row[col], (counts.get(row[col]) || 0) + 1);
          }
        }
        let topVal: any = null;
        let topCount = 0;
        for (const [val, c] of counts.entries()) {
          if (c > topCount) {
            topCount = c;
            topVal = val;
          }
        }
        if (topVal !== null) {
          for (const row of currentData) {
            if (this.isNull(row[col])) {
              row[col] = topVal;
            }
          }
        }
      } else if (estrategia === "ffill") {
        let lastVal: any = null;
        for (const row of currentData) {
          if (!this.isNull(row[col])) {
            lastVal = row[col];
          } else if (lastVal !== null) {
            row[col] = lastVal;
          }
        }
      } else if (estrategia === "bfill") {
        let nextVal: any = null;
        for (let i = currentData.length - 1; i >= 0; i--) {
          const row = currentData[i];
          if (!this.isNull(row[col])) {
            nextVal = row[col];
          } else if (nextVal !== null) {
            row[col] = nextVal;
          }
        }
      }
      this._log(`[NULOS] '${col}' tratado com estrategia '${estrategia}'.`);
    }

    return {
      columns: [...df.columns],
      data: currentData,
    };
  }

  /**
   * Padroniza texto: upper, lower, title, strip, remover_acentos.
   */
  public padronizarTexto(
    df: DataFrame,
    colunas: string[],
    modo: "upper" | "lower" | "title" | "strip" | "remover_acentos" = "upper",
    removerEspacosExtras: boolean = true
  ): DataFrame {
    const newData = df.data.map((r) => ({ ...r }));

    const removeAccents = (str: string) => {
      return str.normalize("NFD").replace(/[\u0300-\u036f]/g, "");
    };

    const toTitleCase = (str: string) => {
      return str.toLowerCase().replace(/(?:^|\s)\S/g, (a) => a.toUpperCase());
    };

    for (const col of colunas) {
      if (!df.columns.includes(col)) continue;
      for (const row of newData) {
        const val = row[col];
        if (val === null || val === undefined) continue;
        let s = String(val);

        if (removerEspacosExtras) {
          s = s.trim().replace(/\s+/g, " ");
        }

        if (modo === "upper") {
          s = s.toUpperCase();
        } else if (modo === "lower") {
          s = s.toLowerCase();
        } else if (modo === "title") {
          s = toTitleCase(s);
        } else if (modo === "remover_acentos") {
          s = removeAccents(s);
        } else if (modo === "strip") {
          s = s.trim();
        }

        row[col] = s;
      }
      this._log(`[TEXTO] '${col}' padronizado (${modo}).`);
    }

    return {
      columns: [...df.columns],
      data: newData,
    };
  }

  /**
   * Corrige tipos de dados das colunas.
   */
  public corrigirTipos(df: DataFrame, mapeamentoTipos: Record<string, string>): DataFrame {
    const newData = df.data.map((r) => ({ ...r }));

    for (const [col, tipo] of Object.entries(mapeamentoTipos)) {
      if (!df.columns.includes(col)) continue;

      for (const row of newData) {
        const val = row[col];
        if (this.isNull(val)) continue;

        if (tipo === "numero" || tipo === "float" || tipo === "int") {
          let s = String(val).trim().replace(/[R$€£%\s]/g, "");
          if (/^-?\d{1,3}(\.\d{3})*,\d+$/.test(s)) {
            s = s.replace(/\./g, "").replace(",", ".");
          } else if (/^-?\d+,\d+$/.test(s)) {
            s = s.replace(",", ".");
          }
          const num = tipo === "int" ? parseInt(s, 10) : parseFloat(s);
          row[col] = isNaN(num) ? null : num;
        } else if (tipo === "texto" || tipo === "string") {
          row[col] = String(val);
        } else if (tipo === "booleano") {
          const s = String(val).trim().toLowerCase();
          row[col] = s === "true" || s === "1" || s === "sim" || s === "s" || s === "verdadeiro";
        }
      }
      this._log(`[TIPO] '${col}' convertido para '${tipo}'.`);
    }

    return {
      columns: [...df.columns],
      data: newData,
    };
  }

  /**
   * Substituição por expressão regular.
   */
  public substituirPorRegex(
    df: DataFrame,
    colunas: string[],
    padrao: string,
    substituicao: string
  ): DataFrame {
    const newData = df.data.map((r) => ({ ...r }));
    const regex = new RegExp(padrao, "g");

    for (const col of colunas) {
      if (!df.columns.includes(col)) continue;
      for (const row of newData) {
        if (!this.isNull(row[col])) {
          row[col] = String(row[col]).replace(regex, substituicao);
        }
      }
      this._log(`[REGEX] '${col}': /${padrao}/ -> '${substituicao}'.`);
    }

    return {
      columns: [...df.columns],
      data: newData,
    };
  }

  /**
   * Tratamento de outliers com IQR.
   */
  public tratarOutliers(
    df: DataFrame,
    colunas: string[],
    fator: number = 1.5,
    acao: "clip" | "remover" | "null" = "clip"
  ): DataFrame {
    let currentData = df.data.map((r) => ({ ...r }));

    for (const col of colunas) {
      if (!df.columns.includes(col)) continue;
      const values: { index: number; num: number }[] = [];
      for (let i = 0; i < currentData.length; i++) {
        const val = currentData[i][col];
        if (!this.isNull(val)) {
          const num = typeof val === "number" ? val : parseFloat(String(val).replace(",", "."));
          if (!isNaN(num)) values.push({ index: i, num });
        }
      }

      if (values.length < 4) continue;
      const sortedNums = values.map((v) => v.num).sort((a, b) => a - b);
      const q1 = sortedNums[Math.floor(sortedNums.length * 0.25)];
      const q3 = sortedNums[Math.floor(sortedNums.length * 0.75)];
      const iqr = q3 - q1;
      const lowerBound = q1 - fator * iqr;
      const upperBound = q3 + fator * iqr;

      if (acao === "clip") {
        for (const v of values) {
          if (v.num < lowerBound) currentData[v.index][col] = lowerBound;
          else if (v.num > upperBound) currentData[v.index][col] = upperBound;
        }
      } else if (acao === "null") {
        for (const v of values) {
          if (v.num < lowerBound || v.num > upperBound) currentData[v.index][col] = null;
        }
      } else if (acao === "remover") {
        currentData = currentData.filter((row) => {
          const val = row[col];
          if (this.isNull(val)) return true;
          const num = typeof val === "number" ? val : parseFloat(String(val).replace(",", "."));
          if (isNaN(num)) return true;
          return num >= lowerBound && num <= upperBound;
        });
      }
      this._log(`[OUTLIERS] '${col}' tratados com IQR (bounds: [${lowerBound.toFixed(2)}, ${upperBound.toFixed(2)}]).`);
    }

    return {
      columns: [...df.columns],
      data: currentData,
    };
  }
}
